### DISCUSSION TITLE
C# - 403 with GitHub API request

### DISCUSSION THREAD

**Discussion by ArunVenkat1994** [community/community#58518] [Apps, API and Webhooks]
### Select Topic Area

General

### Body

I'm having an a 403 response when I try to get the content of a repo on GitHub.

Using a curl command " curl -L -H "Accept: application/vnd.github+json" -H "Authorization: Bearer <my_token>" -H "X-GitHub-Api-Version: 2022-11-28" [https://api.github.com/repos/Arcadix/Product.Fusion/contents"](https://api.github.com/repos/Arcadix/Product.Fusion/contents%22) gives me the right content, but using C# I get a 403 error. Here is the code:

var httpClient = new HttpClient();
httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", "token123"); //token123 is replaced by my token of course
httpClient.DefaultRequestHeaders.Add("Accept", "application/vnd.github+json");
httpClient.DefaultRequestHeaders.Add("X-GitHub-Api-Version", "2022-11-28");
var contentsUrl = $"https://api.github.com/repos/Arcadix/Product.Fusion/contents";
var contentsJson = httpClient.GetStringAsync(contentsUrl).Result;
Console.WriteLine(contentsJson);

**Comment by leedavi** [comment]
Hi, Did you ever find an answer?

I have the same problem and am finding it impossible to understand why.

curl -L \  -H "Accept: application/vnd.github+json" \  -H "Authorization: Bearer XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX" \   https://api.github.com/user/orgs


CURL works correctly....


        private static void ListOrganizations(string path)
        {
            var url = "https://api.github.com/user/orgs";
            var githubToken = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX";
            using (var client = new System.Net.Http.HttpClient())
            {
                client.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", githubToken);                
                var acceptitem = new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/vnd.github+json");
                client.DefaultRequestHeaders.Accept.Add(acceptitem);
            }

        }


Fails with 403.

**Comment by leedavi** [comment]
This madness works...

        public static string ExecuteCurl(string curlCommand)
        {
            if (string.IsNullOrEmpty(curlCommand)) return "";
            curlCommand = curlCommand.Trim();

            // remove the curl keworkd
            if (curlCommand.StartsWith("curl"))
            {
                curlCommand = curlCommand.Substring("curl".Length).Trim();
            }
            curlCommand = curlCommand.Replace("--compressed", "");


            using (var proc = new Process())
            {
                var startInfo = new ProcessStartInfo();
                startInfo.FileName = "curl.exe";
                startInfo.Arguments = curlCommand.ToString();
                startInfo.UseShellExecute = false;
                startInfo.RedirectStandardOutput = true;
                startInfo.RedirectStandardError = true;
                startInfo.CreateNoWindow = true;
                startInfo.WorkingDirectory = Environment.SystemDirectory;

                proc.StartInfo = startInfo;
                proc.Start();
                return proc.StandardOutput.ReadToEnd();
            }
        }


But you need curl installed on the machine, so it's useless if you want to deploy it.
